src/sm.c

```diff
@@ -0,0 +1,331 @@
+    volatile LONG e_done; /*varible that is signalled when e == n-1 (so a barrier is signaled that it might be allowed to proceed with execution)*/
+#ifdef _MSC_VER
+/*warning C4200: nonstandard extension used: zero-sized array in struct/union : looks very standard in C99 and it is called flexible array. Documentation-wise is a flexible array, but called "unsized" in Microsoft's docs*/ /*https://msdn.microsoft.com/en-us/library/b6fae073.aspx*/
+#pragma warning(disable:4200)
+#endif
+
+    char name[]; /*used in printing "who this is"*/
+}SM_HANDLE_DATA;
+
+static const char NO_NAME[] = "NO_NAME";
+
+MU_DEFINE_ENUM_STRINGS(SM_RESULT, SM_RESULT_VALUES);
+
+SM_HANDLE sm_create(const char* name)
+{
+    SM_HANDLE result;
+    
+    /*Codes_SRS_SM_02_001: [ If name is NULL then sm_create shall behave as if name was "NO_NAME". ]*/
+    if (name == NULL)
+    {
+        name = NO_NAME; 
+    }
+    
+    size_t flexSize = strlen(name) + 1;
+
+    /*Codes_SRS_SM_02_002: [ sm_create shall allocate memory for the instance. ]*/
+    result = malloc(sizeof(SM_HANDLE_DATA) + flexSize);
+
+    if (result == NULL)
+    {
+        /*Codes_SRS_SM_02_004: [ If there are any failures then sm_create shall fail and return NULL. ]*/
+        LogError("SM name=%s, failure in malloc(sizeof(SM_HANDLE_DATA)=%zu);",
+            MU_P_OR_NULL(name), sizeof(SM_HANDLE_DATA));
+        /*return as is*/
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_003: [ sm_create shall set b_now to -1, n to 0, and e to 0 succeed and return a non-NULL value. ]*/
+        (void)InterlockedExchange64(&result->b_now, -1);
+        (void)InterlockedExchange64(&result->n, 0);
+        (void)InterlockedExchange64(&result->e, 0);
+        (void)InterlockedExchange(&result->e_done, 0);
+        (void)memcpy(result->name, name, flexSize);
+        /*return as is*/
+    }
+    return result;
+}
+
+
+void sm_destroy(SM_HANDLE sm)
+{
+    /*Codes_SRS_SM_02_005: [ If sm is NULL then sm_destroy shall return. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_006: [ sm_destroy shall free all used resources. ]*/
+        free(sm);
+    }
+}
+
+int sm_open_begin(SM_HANDLE sm)
+{
+    SM_RESULT result;
+    if (sm == NULL)
+    {
+        /*Codes_SRS_SM_02_007: [ If sm is NULL then sm_open_begin shall fail and return SM_ERROR. ]*/
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+        result = SM_ERROR;
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_008: [ If b_now is not -1 then sm_open_begin shall fail and return SM_EXEC_REFUSED. ]*/
+        /*Codes_SRS_SM_02_009: [ sm_open_begin shall set b_now to 0, succeed and return SM_EXEC_GRANTED. ]*/
+        LONG64 b_now = InterlockedCompareExchange64(&sm->b_now, 0, -1);
+        if (b_now != -1)
+        {
+            /*Codes_SRS_SM_02_012: [ If sm_open_end doesn't follow a call to sm_open_begin then sm_open_end shall return. ]*/
```

Comment:
>then sm_open_end shall return. [](start = 91, length = 30)

This spec is misplaced, it pertains to sm_open_end.
I noticed it looking at sm_open_end to see if we check in anyway that sm_open_end changes the state of the module only if it is after sm_open_begin.